pkg/dnsclient/dns.go

```diff
@@ -0,0 +1,347 @@
+)
+
+// exchangeOne performs a single DNS exchange with a randomly chosen server
+// out of the server list, returning the response, time, and error (if any).
+// We assume that the upstream resolver requests and validates DNSSEC records
+// itself.
+func (c *Client) exchangeOne(ctx context.Context, hostname string, qtype uint16) (resp *dns.Msg, err error) {
+	m := new(dns.Msg)
+	// Set question type
+	m.SetQuestion(dns.Fqdn(hostname), qtype)
+	// Set the AD bit in the query header so that the resolver knows that
+	// we are interested in this bit in the response header. If this isn't
+	// set the AD bit in the response is useless (RFC 6840 Section 5.7).
+	// This has no security implications, it simply allows us to gather
+	// metrics about the percentage of responses that are secured with
+	// DNSSEC.
+	m.AuthenticatedData = true
+	// Tell the resolver that we're willing to receive responses up to 4096 bytes.
+	// This happens sometimes when there are a very large number of CAA records
+	// present.
+	m.SetEdns0(4096, false)
+
+	if len(c.servers) < 1 {
+		return nil, errors.Errorf("not configured with at least one DNS Server")
+	}
+
+	// Randomly pick a server
+	chosenServerIndex := rand.Intn(len(c.servers))
+	chosenServer := c.servers[chosenServerIndex]
+
+	tries := 1
+	qtypeStr := dns.TypeToString[qtype]
+	started := time.Now().UTC()
+	// Publish metrics
+	defer func() {
+		result := "failed"
+		if resp != nil {
+			result = dns.RcodeToString[resp.Rcode]
+		}
+
+		metrics.MeasureSince(
+			metricsKeyForTotalLookup,
+			started,
+			metrics.Tag{Name: "qtype", Value: qtypeStr},
+			metrics.Tag{Name: "result", Value: result},
+			metrics.Tag{Name: "retries", Value: strconv.Itoa(tries)},
+			metrics.Tag{Name: "resolver", Value: chosenServer},
+		)
+	}()
+
+	type dnsResp struct {
+		m   *dns.Msg
+		err error
+	}
+
+	for {
+		ch := make(chan dnsResp, 1)
+
+		go func() {
+			logger.Tracef("hostname=%q, type=%v, server=%s", hostname, qtypeStr, chosenServer)
+
+			started := time.Now().UTC()
+			rsp, _, err := c.dnsClient.Exchange(m, chosenServer)
+			result := "failed"
+			if rsp != nil {
+				result = dns.RcodeToString[rsp.Rcode]
+			}
+			metrics.MeasureSince(
+				metricsKeyForQuery,
+				started,
+				metrics.Tag{Name: "qtype", Value: qtypeStr},
+				metrics.Tag{Name: "result", Value: result},
+				metrics.Tag{Name: "resolver", Value: chosenServer},
+			)
+
+			ch <- dnsResp{m: rsp, err: err}
+		}()
+		select {
+		case <-ctx.Done():
+			reason := "unknown"
+			if ctx.Err() == context.DeadlineExceeded {
+				reason = "deadline"
+			} else if ctx.Err() == context.Canceled {
+				reason = "canceled"
+			}
+			metrics.IncrCounter(
+				metricsKeyForTimeoutCounter,
+				1,
+				metrics.Tag{Name: "qtype", Value: qtypeStr},
+				metrics.Tag{Name: "reason", Value: reason},
+				metrics.Tag{Name: "resolver", Value: chosenServer},
+			)
+
+			err = ctx.Err()
+			logger.Errorf("hostname=%q, type=%v, err=[%s]", hostname, qtypeStr, err.Error())
```

Comment:
This is a strange line and potentially can cause panic if err is nil. Instead maybe the below one is better.

```
if err != nil{
   logger.Errorf("hostname=%q, type=%v, err=[%s]", hostname, qtypeStr, err.Error())
} else {
   logger.Errorf("hostname=%q, type=%v, err=context has finished without any specific error.")
}
```

Reply:
fixed